test/unit/test_type.c

```diff
@@ -455,6 +455,63 @@ static bool test_const_types(void) {
 	mu_end;
 }
 
+static char *array = "int a[65][5][0]";
+static char *array_exp1 = "int [65][5][0]";
+static char *array_ptr = "int * const *a[][][][9]";
+static char *array_ptr_exp1 = "int * const *[0][0][0][9]";
+static char *array_ptr_exp2 = "int * const *a[0][0][0][9]";
+static char *struct_array_ptr = "struct alb { const char *b; int * const *a[][][][9]; }";
+static char *struct_array_ptr_exp1 = "struct alb";
+static char *struct_array_ptr_exp2 = "struct alb { const char *b; int * const *a[0][0][0][9]; }";
+static char *struct_array_ptr_exp3 = "struct alb a";
+
+static bool test_type_as_string(void) {
+	RzTypeDB *typedb = rz_type_db_new();
+	mu_assert_notnull(typedb, "Couldn't create new RzTypeDB");
+	mu_assert_notnull(typedb->types, "Couldn't create new types hashtable");
+	const char *dir_prefix = rz_sys_prefix(NULL);
+	rz_type_db_init(typedb, dir_prefix, "x86", 64, "linux");
+
+	char *error_msg = NULL;
+	RzType *ttype = rz_type_parse_string_single(typedb->parser, array, &error_msg);
+	mu_assert_notnull(ttype, "array type parse successfull");
+	mu_assert_true(ttype->kind == RZ_TYPE_KIND_ARRAY, "is array");
+
+	char *array_str1 = rz_type_as_string(typedb, ttype);
+	char *array_str2 = rz_type_declaration_as_string(typedb, ttype);
+	char *array_str3 = rz_type_identifier_declaration_as_string(typedb, ttype, "a");
+	mu_assert_streq_free(array_str1, array_exp1, "rz_type_as_string");
+	mu_assert_streq_free(array_str2, array_exp1, "rz_type_declaration_as_string");
+	mu_assert_streq_free(array_str3, array, "rz_type_identifier_declaration_as_string");
+	rz_type_free(ttype);
+
+	ttype = rz_type_parse_string_single(typedb->parser, array_ptr, &error_msg);
+	mu_assert_notnull(ttype, "array type parse successfull");
+	mu_assert_true(ttype->kind == RZ_TYPE_KIND_ARRAY, "is array");
+
+	array_str1 = rz_type_as_string(typedb, ttype);
+	array_str2 = rz_type_declaration_as_string(typedb, ttype);
+	array_str3 = rz_type_identifier_declaration_as_string(typedb, ttype, "a");
+	mu_assert_streq_free(array_str1, array_ptr_exp1, "rz_type_as_string");
+	mu_assert_streq_free(array_str2, array_ptr_exp1, "rz_type_declaration_as_string");
+	mu_assert_streq_free(array_str3, array_ptr_exp2, "rz_type_identifier_declaration_as_string");
```

Comment:
I think that the "pretty print" function developed by @DMaroo should be able to print, with different options, all these different types of output. What do you think?

However, I'm not sure I do see the difference between `rz_type_as_string` and `rz_type_identifier_declaration_as_string`. Could you explain a bit?

Also, the doc for those functions should be very explicit about what they output and maybe even provide some examples IMO.

Reply:
I explained this before already:
- `rz_type_as_string()` shows pure type without identifier and unfolds it: `struct bla { int a; char foo; }`
- `rz_type_declaration_as_string()` shows pure type without identifier without unfolding it: `struct bla`
- `rz_type_identifier_declaration_as_string()` shows type with identifier: `struct bla foo`. Moreover, it also expands the anonymous structures inline. They all are used in different parts of the code and serve different purpose. We need them all.

Reply:
Ok yeah it's clearer with `struct_array_ptr` test. Please make sure those things are properly captured in the doxygen doc. And by looking at that example actually it seems that `rz_type_as_string` does not unfold the struct, while `rz_type_declaration_as_string` does. Right?

So, to wrap up this is what I understand:
- rz_type_as_string == nothing special
- rz_type_declaration_as_string == UNFOLD
- rz_type_identifier_declaration_as_string == UNFOLD + IDENTIFIER
- `tc` output == MULTILINE (no UNFOLD and no IDENTIFIER, right?)

Another question: do we need to differentiate between UNFOLD and UNFOLD_INNER (meaning that it unfolds also the inner structures, if present)? I'm trying to gather all the requirements to have a single pretty-print function for @DMaroo .

Reply:
Using `rz_type_identifier_declaration_as_string()` is very important since we need to put all pointer things *before* identifier and array things *after* identifier. Thus it's required to pass the identifier *inside* the function during the string formation, it's crucial - we cannot simple concatenate "typestr + identifier".

Reply:
@ret2libc @DMaroo we likely need to have 2 different unfolds as well:
- `UNFOLD_ALL` (unfold all structures/unions that are not typedefs)
- `UNFOLD_ANONYMOUS` (only anonymous structures/unions)
Apart from that right, yes

Reply:
(For the pretty print function) How should cases of self-reference (linked list) or circular references be resolved while unfolding a struct? Should the unfolding be avoided if the type in question is same as the parent type?